docs/tutorials/python/sagas/index.md

```diff
@@ -0,0 +1,586 @@
+Each step in the Workflow has a corresponding compensation step that is executed in reverse order if the Workflow encounters an error.
+
+This ensures that the system is returned to a consistent state, even in the case of partial failures.
+
+Create a new file named `saga_workflows.py`.
+This file will define your Workflow, which is responsible for executing your Activities in the correct order and handling compensation if necessary.
+
+First, import the necessary modules:
+
+```python
+from datetime import timedelta
+from temporalio import workflow
+from temporalio.common import RetryPolicy
+
+with workflow.unsafe.imports_passed_through():
+    from activities import BookVacationInput, book_car, book_flight, book_hotel
+```
+
+Next, create the `BookWorkflow` class and define the compensation actions, as well as the functions that execute your core logic: `book_car`, `book_hotel`, and `book_flight`.
+
+These executions are wrapped in a `try` and `except` block to handle any exceptions and trigger compensations.
+
+```python
+@workflow.defn
+class BookWorkflow:
+    @workflow.run
+    async def run(self, input: BookVacationInput):
+        compensations = []
+
+        try:
+            # Attempt to book a car
+            compensations.append("undo_book_car")
+            output = await workflow.execute_activity(
+                book_car,
+                input,
+                start_to_close_timeout=timedelta(seconds=10),
+                retry_policy=RetryPolicy(
+                    non_retryable_error_types=["Exception"],
+                ),
+            )
+
+            # Attempt to book a hotel
+            compensations.append("undo_book_hotel")
+            output += " " + await workflow.execute_activity(
+                book_hotel,
+                input,
+                start_to_close_timeout=timedelta(seconds=10),
+                retry_policy=RetryPolicy(
+                    non_retryable_error_types=["Exception"],
+                ),
+            )
+
+            # Attempt to book a flight
+            compensations.append("undo_book_flight")
+            output += " " + await workflow.execute_activity(
+                book_flight,
+                input,
+                start_to_close_timeout=timedelta(seconds=10),
+                retry_policy=RetryPolicy(
+                    initial_interval=timedelta(seconds=1),
+                    maximum_interval=timedelta(seconds=1),
+                    maximum_attempts=input.attempts,
+                    non_retryable_error_types=["Exception"],
+                ),
+            )
+
+            # If all bookings are successful, return the output
+            return output
+        except Exception:
+            # If an error occurs, execute compensations in reverse order
+            for compensation in reversed(compensations):
+                await workflow.execute_activity(
```

Comment:
You should explain what happens when the undo activity fails to confirm that behavior is intentional